### PR TITLE
[Feature](bangc-ops): add new op active_rotated_filter_forward.

### PR DESCRIPTION
**Pull request by DanieeelLiu**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

add new operator active_rotated_filter_forward.

## 2. Modification
new:
bangc-ops/kernels/active_rotated_filter/active_rotated_filter.cpp
bangc-ops/kernels/active_rotated_filter/active_rotated_filter_unionx.mlu
bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/active_rotated_filter_forward/active_rotated_filter_forward.cpp
bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/active_rotated_filter_forward/active_rotated_filter_forward.h
bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/active_rotated_filter_forward/test_case/case0_.prototxt
docs/bangc-docs/design_docs/active_rotated_filter_forward/ORF.png
docs/bangc-docs/design_docs/active_rotated_filter_forward/active_rotated_filter_forward.md
docs/bangc-docs/design_docs/active_rotated_filter_forward/rotate45.png
docs/bangc-docs/design_docs/active_rotated_filter_forward/rotate_degree.png

modified:
bangc-ops/kernel_depends.toml
bangc-ops/mlu_op.h
bangc-ops/mlu_op_kernel.h

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard
Diff3 == 0

#### 3.1.2 Operator Scheme checklist

|     No.      |           Details            |      Check Results       |
|----------------|---------------------------|---------------------|
|        1       |          Supported hardware         |  MLU290 <br>MLU370<br>MLU590|
|        2       |          Job types          |    block <br> U1 <br> U4    |
|        3       |         Layouts            |  NHWC 、NCHW、ARRAY etc    |
|        4       |         Whether multi-dimensions are supported              |                |
|        5       |          Whether element zero is supported             |               |
|        6       |         Data type(half/float)       |         half / float etc           |
|        7      |        Whether there is size limit           |             |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [x] Multidimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md).

### 3.2 Accuracy Test
290:
[----------] Global test environment tear-down
[ SUMMARY  ] Total 87 cases of 1 op(s).
ALL PASSED.
[==========] 87 test cases from 1 test suite ran.
[  PASSED  ] 87 test cases.

370:
[ SUMMARY  ] Total 87 cases of 1 op(s).
ALL PASSED.
[==========] 87 test cases from 1 test suite ran. 
[  PASSED  ] 87 test cases.

590:
[----------] Global test environment tear-down
[ SUMMARY  ] Total 87 cases of 1 op(s).
ALL PASSED.
[==========] 87 test cases from 1 test suite ran. 
[  PASSED  ] 87 test cases.

### 3.4 Summary Analysis
All test cases passed.

### CODE REVIEW

**Review by wushaoqiang2168 — APPROVED** [review]

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -6455,8 +6455,147 @@ mluOpStatus_t MLUOP_WIN_API mluOpGetReduceOpWorkspaceSize(mluOpHandle_t handle,
...
+ * @par Data Layout
+ * - The supported data layouts of \b input, \b indices, and \b output are as follows:
+ *   - input tensor: MLUOP_LAYOUT_ARRAY
+ *   - output tensor: MLUOP_LAYOUT_ARRAY
+ *   - indices tensor: MLUOP_LAYOUT_ARRAY
+ *
+ * @par Scale Limitation
+ * - The \b input is 5D array, and \b indices and \b output are 4D array.
+ * - The dims[2] of \b input should be equal to the power of 2 and less than or
+ * equanl to 128, dims[3] should be equal to 1 or 3, and dims[3] shoule be equal
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6550]: equanl   拼写错误

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -6455,8 +6455,147 @@ mluOpStatus_t MLUOP_WIN_API mluOpGetReduceOpWorkspaceSize(mluOpHandle_t handle,
                                                           const mluOpReduceDescriptor_t reduce_op,
                                                           size_t *workspace_size_inbytes);
 
+// Group:ActiveRotatedFilterForward
+/*!
+ * @brief Returns in \b workspace_size the size of the MLU memory that is used as an extra
+ * workspace to optimize the ActiveRotatedFilterForward operation. The size of the extra
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6461]: ActiveRotatedFilterForward这个是算子的名称吧？是不是这个mluOpActiveRotatedFilterForward

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -6455,8 +6455,147 @@ mluOpStatus_t MLUOP_WIN_API mluOpGetReduceOpWorkspaceSize(mluOpHandle_t handle,
...
+ *
+ */
+mluOpStatus_t MLUOP_WIN_API
+mluOpGetActiveRotatedFilterForwardWorkspaceSize(const mluOpHandle_t handle,
+                                                const mluOpTensorDescriptor_t input_desc,
+                                                size_t *workspace_size);
+
+// Group:ActiveRotatedFilterForward
+/*!
+ * @brief Rotate \b input according to \b indices . This function encodes
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6498]: Rotates   句号前多了一个空格

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -6455,8 +6455,147 @@ mluOpStatus_t MLUOP_WIN_API mluOpGetReduceOpWorkspaceSize(mluOpHandle_t handle,
...
+ *   - output tensor: MLUOP_LAYOUT_ARRAY
+ *   - indices tensor: MLUOP_LAYOUT_ARRAY
+ *
+ * @par Scale Limitation
+ * - The \b input is 5D array, and \b indices and \b output are 4D array.
+ * - The dims[2] of \b input should be equal to the power of 2 and less than or
+ * equanl to 128, dims[3] should be equal to 1 or 3, and dims[3] shoule be equal
+ * to dims[4].
+ * - The dims[0] of \b indices should be equal to \b input dims[2], and dims[1]
+ * and dims[2] of \b indices should be equanl to dims[3] and dims[4] of \b input
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6553]: equanl 拼写错误

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -6455,8 +6455,147 @@ mluOpStatus_t MLUOP_WIN_API mluOpGetReduceOpWorkspaceSize(mluOpHandle_t handle,
...
+ * - The dims[2] and dims[3] of \b output should be equal to dims[3] and dims[4]
+ * of \b input respectively.
+ *
+ * @par API Dependency
+ * - Before calling this function, you need to call
+ * ::mluOpGetActiveRotatedFilterForwardWorkspaceSize to get the extra space size
+ * needed in ActiveRotatedFilterForward operation.
+ *
+ * @par Performance Optimization
+ * - None.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6569]: 这个字段删了吧

**Review by GraceGuanH — COMMENTED** [review]
Please check the wiki guide and update the .h file.

File: bangc-ops/mlu_op.h (hunk 1 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h (hunk 2 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h (hunk 3 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h (hunk 4 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h (hunk 5 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

**Review by GraceGuanH — APPROVED** [review]

**Review by WangYuan-cn — APPROVED** [review]

**Pull request merged**